ipld/resolver/src/service.rs

```diff
@@ -1,17 +1,431 @@
+                if let Err(e) = self.membership_mut().add_provided_subnet(id) {
+                    error!("error adding provided subnet: {e}")
+                }
+            }
+            Request::RemoveProvidedSubnet(id) => {
+                if let Err(e) = self.membership_mut().remove_provided_subnet(id) {
+                    error!("error removing provided subnet: {e}")
+                }
+            }
+            Request::PinSubnet(id) => self.membership_mut().pin_subnet(id),
+            Request::UnpinSubnet(id) => self.membership_mut().unpin_subnet(&id),
+
+            Request::Resolve(cid, subnet_id, response_channel) => {
+                let peers = self.membership_mut().providers_of_subnet(&subnet_id);
+                if peers.is_empty() {
+                    send_resolve_result(response_channel, Err(anyhow!(NoKnownPeers(subnet_id))));
+                } else {
+                    let (connected, known) = peers
+                        .into_iter()
+                        .partition::<Vec<_>, _>(|id| self.swarm.is_connected(id));
+
+                    let (peers, fallback) = if connected.is_empty() {
+                        (known, vec![])
+                    } else {
+                        // Use just the connected ones, however many there are.
+                        // Alternatively we could take the first N combined.
+                        (connected, known)
+                    };
+
+                    let query = Query {
+                        cid,
+                        subnet_id,
+                        response_channel,
+                        fallback_peer_ids: fallback,
+                    };
+
+                    let query_id = self.content_mut().resolve(cid, peers);
+
+                    self.queries.insert(query_id, query);
+                }
+            }
+        }
+    }
+
+    // The following are helper functions because Rust Analyzer has trouble with recognising that `swarm.behaviour_mut()` is a legal call.
+
+    fn discovery_mut(&mut self) -> &mut behaviour::discovery::Behaviour {
+        self.swarm.behaviour_mut().discovery_mut()
+    }
+    fn membership_mut(&mut self) -> &mut behaviour::membership::Behaviour {
+        self.swarm.behaviour_mut().membership_mut()
+    }
+    fn content_mut(&mut self) -> &mut behaviour::content::Behaviour<P> {
+        self.swarm.behaviour_mut().content_mut()
+    }
+}
+
+/// Respond to the sender of the query, if they are still listening.
+fn send_resolve_result(tx: Sender<ResolveResult>, res: ResolveResult) {
+    if tx.send(res).is_err() {
+        error!("error sending resolve result; listener closed")
+    }
+}
+
+/// Builds the transport stack that libp2p will communicate over.
+///
+/// Based on the equivalent in Forest.
+pub fn build_transport(local_key: Keypair) -> Boxed<(PeerId, StreamMuxerBox)> {
+    let tcp_transport =
+        || libp2p::tcp::tokio::Transport::new(libp2p::tcp::Config::new().nodelay(true));
```

Comment:
I don't know if it is already supported in rust-libp2p, but I would love if we could switch to quic in the future and only fallback to tcp if not supported :)

Reply:
I will check. FWIW in #64 I factor out transport creation, so we can switch easily.